enarx-keep-sgx-shim/src/handler.rs

```diff
@@ -288,25 +288,31 @@ impl<'a> Handler<'a> {
         let buf = self.aex.gpr.rsi.raw() as *const u8;
         let size = self.aex.gpr.rdx.raw();
 
-        // Allocate some unencrypted memory.
-        let map = match self.ualloc(size) {
-            Err(errno) => return errno as u64,
-            Ok(map) => map,
-        };
+        // Copy the encrypted input into unencrypted memory.
+        let input = unsafe { from_raw_parts(buf, size as usize) };
+        let ubuf = unsafe { from_raw_parts_mut(&self.block.buf as *const _ as _, size as usize) };
+        ubuf.copy_from_slice(input);
 
-        unsafe {
-            // Copy the encrypted input into unencrypted memory.
-            copy_nonoverlapping(buf, map, size as _);
-
-            // Do the syscall; replace encrypted memory with unencrypted memory.
-            let ret = self.syscall(libc::SYS_write as u64, fd, map as _, size, 0, 0, 0);
-            self.ufree(map, size);
+        // Do the syscall; replace encrypted memory with unencrypted memory.
+        let res = unsafe {
+            self.proxy(Request::new(
+                libc::SYS_write as usize,
+                &[
+                    fd.into(),
+                    (&self.block.buf as *const _ as usize).into(),
+                    size.into(),
+                ],
+            ))
+        };
 
-            if ret <= ERRNO_BASE && ret > size {
-                self.attacked();
+        match res {
+            Ok(res) => {
+                if res[0].raw() as u64 > size {
```

Comment:
```suggestion
                if res[0].into() > size {
```

Reply:
The compiler couldn't infer a type for this, so I have it as `.raw()` (but not a `u64` anymore since `size` is a `usize` now).